eng/WpfArcadeSdk/tools/InjectModuleInitializer/AddModuleConstructorTask.cs

```diff
@@ -1,80 +1,80 @@
-                        return false;
-                    }
-
-                    if (Regex.Match(sourceIL, @"\.assembly extern mscorlib", RegexOptions.Singleline) == Match.Empty)
-                    {
-                        outputIL.WriteLine(msCorLibAssemblySectionIL);
-                    }
-
-                    if (Regex.Match(sourceIL, @"\.class private auto ansi '\<Module\>'.+?\.method private hidebysig specialname rtspecialname static void \.cctor \(\) cil managed ", RegexOptions.Singleline) == Match.Empty)
-                    {
-                        outputIL.WriteLine(moduleConstructorIL);
-                    }
-                    else
-                    {
-                        Log.LogError("Cannot insert a module initializer into an assembly that already contains one.");
-                        return false;
-                    }
-
-                    outputIL.WriteLine(sourceIL);
-                }
-
-                return true;
-            }
-            catch (Exception e)
-            {
-                Log.LogError(e.ToString() + Environment.NewLine + e.StackTrace);
-                return false;
-            }
-        }
-    }
-}
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Text.RegularExpressions;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace WpfArcadeSdk.Build.Tasks
+{
+    /// <summary>
+    /// Append the supplied module constructor to the supplied IL file.
+    /// Using this requires that the DLL has implemented a static class named
+    /// ModuleInitializer with a static void Initialize function.  This class
+    /// is called from the module initializer.
+    /// </summary>
+    public class AddModuleConstructorTask : Task
+    {
+        /// <summary>
+        /// The original IL file to manipulate
+        /// </summary>
+        [Required]
+        public string ILFile { get; set; }
+
+        /// <summary>
+        /// Contains the IL section referencing mscorlib.
+        /// This is inserted if not already present.
+        /// </summary>
+        [Required]
+        public string MsCorLibAssemblySectionIL { get; set; }
+
+        /// <summary>
+        /// Contains the IL of the module constructor
+        /// </summary>
+        [Required]
+        public string ModuleConstructorIL { get; set; }
+
+        public override bool Execute()
+        {
+            try
+            {
+                var moduleConstructorIL = File.ReadAllText(ModuleConstructorIL);
+                var msCorLibAssemblySectionIL = File.ReadAllText(MsCorLibAssemblySectionIL);
+                var sourceIL = File.ReadAllText(ILFile);
+
+                using (var outputIL = new StreamWriter(File.Open(ILFile, FileMode.Truncate)))
+                {
+                    if (Regex.Match(sourceIL, @"\.assembly extern mscorlib", RegexOptions.Singleline) == Match.Empty)

```

Comment:
If sourceIL already contains mscorlib reference, shouldn't you fail now?   If I understand what you've told me, that situation will produce IL where the module initializer precedes the mscorlib reference, which leads ilasm to fail silently (by not including the module init).  So catching this now when we can complain loudly seems good.

Reply:
Yeah, when we first did this we wanted to keep it general.  However, given the failure mode is silent, we should be as loud as possible when things seem incorrect.

Reply:
I decided just to make it work by replacing the mscorlib reference when we find it with the combined set of what we mean to inject (assembly reference included).  If we do not find it, we inject both from scratch at the beginning.